pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -215,130 +208,130 @@ func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
-			klogV.InfoS("No update required", "wrapper", wrapper.TypeName(), "type", configurationType, "name", name,
-				"diff", cmp.Diff(current, bootstrap))
+			klogV.InfoS("No update required", "wrapper", bootstrap.typeName(), "type", configurationType, "name", name,
+				"diff", cmp.Diff(wah.getHave(), wah.getWant()))
 		}
 		return nil
 	}
 
-	if _, err = wrapper.Update(newObject); err == nil {
-		klog.V(2).Infof("Updated the %s type=%s name=%q diff: %s", wrapper.TypeName(), configurationType, name, cmp.Diff(current, newObject))
+	if err = newObject.update(ctx); err == nil {
+		klog.V(2).Infof("Updated the %s type=%s name=%q diff: %s", bootstrap.typeName(), configurationType, name, cmp.Diff(wah.getHave(), wah.getWant()))
 		return nil
 	}
 
 	if apierrors.IsConflict(err) {
-		klog.V(2).InfoS(fmt.Sprintf("Something updated the %s concurrently, I will check its spec later", wrapper.TypeName()), "type", configurationType, "name", name)
+		klog.V(2).InfoS(fmt.Sprintf("Something updated the %s concurrently, I will check its spec later", bootstrap.typeName()), "type", configurationType, "name", name)
 		return nil
 	}
 
-	return fmt.Errorf("failed to update the %s, will retry later type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	return fmt.Errorf("failed to update the %s, will retry later type=%s name=%q error=%w", bootstrap.typeName(), configurationType, name, err)
 }
 
-// removeAutoUpdateEnabledConfiguration makes an attempt to remove the given
-// configuration object if automatic update of the spec is enabled for this object.
-func removeAutoUpdateEnabledConfiguration(wrapper configurationWrapper, name string) error {
-	current, err := wrapper.Get(name)
+// RemoveUnwantedObjects attempts to delete the configuration objects
+// that exist, are annotated `apf.kubernetes.io/autoupdate-spec=true`, and do not
+// have a name in the given set.  A refusal due to concurrent update is logged
+// and not considered an error; the object will be reconsidered later.
+func RemoveUnwantedObjects(ctx context.Context, boots BootstrapObjects) error {
+	current, err := boots.getExistingObjects()
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil
-		}
-
-		return fmt.Errorf("failed to retrieve the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
-	}
-
-	value := current.GetAnnotations()[flowcontrolv1beta3.AutoUpdateAnnotationKey]
-	autoUpdate, err := strconv.ParseBool(value)
-	if err != nil {
-		klog.ErrorS(err, fmt.Sprintf("Skipping deletion of the %s", wrapper.TypeName()), "name", name)
-
-		// This may need manual intervention, in case the annotation value is malformed,
-		// so don't return an error, that might trigger futile retry loop.
-		return nil
-	}
-	if !autoUpdate {
-		klog.V(5).InfoS(fmt.Sprintf("Skipping deletion of the %s", wrapper.TypeName()), "name", name)
-		return nil
+		klog.ErrorS(err, "Failed to list existing APF configuration objects", "type", boots.typeName())
```

Comment:
Shouldn't we handle NotFound error differently, as before?

Reply:
(1) Does a lister ever return a NotFound error?
(2) If the lister ever actually did return a NotFound error, there is no more work to do about removing unwanted objects.
(3) I think the behavior here matches the behavior of what is currently merged.

Reply:
> (1) Does a lister ever return a NotFound error?

If the object is not in the cache, it will return NotFound.
But I agree not for list.

> (3) I think the behavior here matches the behavior of what is currently merged.

You're proceeding with further processing in case of non-nil error.
But in that case, current might be nil.
This should work fine, but I would feel safer if we simply return here.

Reply:
Oh, I see.  Actually, on further examination, I see that returning an error in this case is what matches the current behavior.  Fixed in the latest revision.